docs/pkg/client/events.md

```diff
@@ -0,0 +1,376 @@
+# Package `pocket/pkg/client/events` <!-- omit in toc -->
+
+> An event query package for interfacing with [CometBFT](https://cometbft.com/)
+> and the [Cosmos SDK](https://v1.cosmos.network/sdk), facilitating subscriptions
+> to chain event messages.
+
+<!-- toc -->
+
+- [Overview](#overview)
+- [Architecture Diagrams](#architecture-diagrams)
+  - [Components](#components)
+  - [Subscriptions](#subscriptions)
+- [Installation](#installation)
+- [Features](#features)
+- [Usage (`EventsQueryClient`)](#usage-eventsqueryclient)
+  - [Basic Example](#basic-example)
+  - [Advanced Usage](#advanced-usage)
+  - [Configuration](#configuration)
+- [Usage (`EventsReplayClient`)](#usage-eventsreplayclient)
+  - [Basic Usage](#basic-usage)
+  - [Advanced Usage](#advanced-usage-1)
+- [Best Practices](#best-practices)
+- [FAQ](#faq)
+  - [Why use `events` over directly using Gorilla WebSockets?](#why-use-events-over-directly-using-gorilla-websockets)
+  - [How can I use a different connection mechanism other than WebSockets?](#how-can-i-use-a-different-connection-mechanism-other-than-websockets)
+  - [Why use the `EventsReplayClient` over directly maintaining an `EventsQueryClient`?](#why-use-the-eventsreplayclient-over-directly-maintaining-an-eventsqueryclient)
+
+<!-- tocstop -->
+
+## Overview
+
+The `events` package provides a client interface to subscribe to chain event
+messages. It abstracts the underlying connection mechanisms and offers a clear
+and easy-to-use way to get events from the chain. Highlights:
+
+- Offers subscription to chain event messages matching a given query.
+- Uses the Gorilla WebSockets package for underlying connection operations.
+- Provides a modular structure with interfaces allowing for mock implementations
+  and testing.
+- Offers considerations for potential improvements and replacements, such as
+  integration with the cometbft RPC client.
+- Offers a generic client to decode on chain event bytes into the desired event
+  type
+
+## Architecture Diagrams
+
+### Components
+
+```mermaid
+---
+title: Component Diagram Legend
+---
+
+flowchart
+
+  a[Component A]
+  b[Component B]
+  c[Component C]
+  d[Component D]
+
+  a --"A uses B via B#MethodName()"--> b
+  a =="A returns C from A#MethodName()"==> c
+  b -."A uses D via network IO".-> d
+```
+
+```mermaid
+---
+title: EventsQueryClient Components
+---
+
+flowchart
+
+  subgraph comet[Cometbft Node]
+    subgraph rpc[JSON-RPC]
+      sub[subscribe endpoint]
+    end
+  end
+
+  subgraph eqc[EventsQueryClient]
+    q1_eb[EventsBytesObservable]
+    q1_conn[Connection]
+    q1_dial[Dialer]
+  end
+
+  q1_obsvr1[Observer 1]
+  q1_obsvr2[Observer 2]
+
+
+  q1_obsvr1 --"#Subscribe()"--> q1_eb
+  q1_obsvr2 --"#Subscribe()"--> q1_eb
+
+
+  q1_dial =="#DialContext()"==> q1_conn
+  q1_eb --"#Receive()"--> q1_conn
+
+  q1_conn -.-> sub
+```
+
+```mermaid
+---
+title: EventsReplayClient Components
+---
```

Comment:
I'm still thinking about this flowchart but the first thing I notice is that it's upside down; take a look at the legend one more time.

Reply:
I'm noticing that `EventsQueryClient` is described here in terms of inert sub-components. Whereas, in `EventsQueryClient Components`, it's (already) described using sub-components which actually associate with others.

Reply:
`Publish Events Factory` seems like it should be named `retryPublishEventsFactory`. Additionally, it's a a function represented as a rectangle, but rectangles represent components in the legend (and rest of the diagram). 

In other places I've used a rounded rectangle to represent functions or methods. Regardless of how that "kind" of thing is represented, it should be:

1. Distinct from other "kinds" which are already defined (for other purposes; i.e. component, usage associations, etc.).
2. Used consistently for that "kind", currently `retry.OnError` is depicted as a usage association. It also needs to change and should be represented the same way as `retryPublishEventsFactory` (assuming  both are still represented in the next iteration).
3. Sufficiently described in the legend to be unambiguous (i.e. hard to misinterpret).

Reply:
FWIW the legend also doesn't currently define rules around encapsulation. That could be taken as an opportunity either to add some and express them in here, or to reconsider representing encapsulation so strictly (e.g. `EventsQueryClient`/`EventsReplayClient`). I don't have a strong opinion here but wanted to share some additional perspective.

Reply:
I'm wondering if it makes sense for either `retry.OnError` or `retryPublishEventsFactory`  to be in this diagram. This diagram should be about the relationship betwen the components that are necessary for this unit to function. 

I think it makes more sense for `retry.OnError` and `retryPublishEventsFactory` to be central in some other diagram about the retry loop; perhaps a flowchart or state diagram showing the ways in which the client can get disconnected, how it recovers, and ultimately aborts and errors. There should be an analogous diagram for the `EventsQueryClient` as it does a similar thing IIRC. Perhaps they can be represented in the same diagram, but it's okay if not.

Reply:
I don't think the `#EventBytes` association is accurately represented as network I/O.

Reply:
> I'm wondering if it makes sense for either `retry.OnError` or `retryPublishEventsFactory` to be in this diagram. This diagram should be about the relationship betwen the components that are necessary for this unit to function.

Similar feedback on `DecoderFunction`. It probably make sense to have a diagram that focuses on data transformation through `EventsReplayClient`, which `DecoderFunction` would be a critical part of. Also, is there a name that's more consistent with the code or have I just not read that part yet?